compiler/luci/pass/src/FuseMulWithConvPass.cpp

```diff
@@ -0,0 +1,213 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "luci/Pass/FuseMulWithConvPass.h"
+
+#include "helpers/NodeFiller.h"
+
+#include <luci/IR/CircleNodes.h>
+#include <luci/Profile/CircleNodeOrigin.h>
+#include <luci/Service/CircleNodeClone.h>
+#include <luci/Service/Nodes/CircleConst.h>
+
+namespace
+{
+
+#define RETURN_FALSE_UNLESS(cond) \
+  if (not(cond))                  \
+    return false;
+
+inline uint32_t cal_offset(const luci::CircleConst *node, const std::vector<uint32_t> &indices)
+{
+  // sanity check for node's rank
+  assert(node != nullptr && node->rank() == 4);
+
+  // sanity check for indices
+  assert(indices.size() == 4);
+
+  return indices[0] * node->dim(1).value() * node->dim(2).value() * node->dim(3).value() +
+         indices[1] * node->dim(2).value() * node->dim(3).value() +
+         indices[2] * node->dim(3).value() + indices[3];
+}
+
+/**
+ *  Fuse Mul with Conv if possible
+ *
+ *  NOTE: In case mul is channewise constant, we can try to merge mul with nconv,
+ *
+ *  BEFORE
+ *                    |
+ *              [CircleConv2D] (no activation)
+ *                    |
+ *                  [Mul] (channel-wise/scalar constant)
+ *                    |
+ *
+ *  AFTER
+ *                    |
+ *            [CircleConv2D] (with updated kernels, bias, and activation)
+ *                    |
+ *
+ */
+
+bool fuse_mul_with_conv(luci::CircleMul *mul)
+{
+  // sanity check
+  RETURN_FALSE_UNLESS(mul->dtype() == loco::DataType::FLOAT32);
+
+  luci::CircleConst *const_mul_operand = nullptr;
+  luci::CircleConv2D *conv = nullptr;
+  RETURN_FALSE_UNLESS(luci::fill(&const_mul_operand, &conv).with_commutative_args_of(mul));
+
+  // sanity check
+  RETURN_FALSE_UNLESS(conv->dtype() == loco::DataType::FLOAT32 &&
+                      const_mul_operand->dtype() == loco::DataType::FLOAT32);
+
+  //  NOTE for general activation function: S * Act(A * B) != Act(A*(SB))
+  if (conv->fusedActivationFunction() != luci::FusedActFunc::NONE)
+    return false;
+
+  // check that const_mul_operand is channel-wise constant or just a scalar
+  RETURN_FALSE_UNLESS(const_mul_operand->rank() == 4 || const_mul_operand->rank() == 1 ||
+                      const_mul_operand->rank() == 0);
+
+  std::vector<float> mul_values;
+  if (const_mul_operand->rank() == 4)
+  {
+    // check channel-wise broadcasting
+    RETURN_FALSE_UNLESS(const_mul_operand->dim(0).value() == 1 &&
+                        const_mul_operand->dim(1).value() == 1 &&
+                        const_mul_operand->dim(2).value() == 1);
+
+    mul_values.resize(const_mul_operand->size<loco::DataType::FLOAT32>());
+    for (uint32_t idx = 0; idx < mul_values.size(); idx++)
+    {
+      mul_values[idx] = const_mul_operand->at<loco::DataType::FLOAT32>(idx);
+    }
```

Comment:
This code is repeated. Can you extract out?

Reply:
Ahh. Ok.

Reply:
done